x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/apikey/TransportUpdateApiKeyAction.java

```diff
@@ -0,0 +1,80 @@
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.action.apikey;
+
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xcontent.NamedXContentRegistry;
+import org.elasticsearch.xpack.core.security.SecurityContext;
+import org.elasticsearch.xpack.core.security.action.apikey.UpdateApiKeyAction;
+import org.elasticsearch.xpack.core.security.action.apikey.UpdateApiKeyRequest;
+import org.elasticsearch.xpack.core.security.action.apikey.UpdateApiKeyResponse;
+import org.elasticsearch.xpack.core.security.authz.RoleDescriptor;
+import org.elasticsearch.xpack.core.security.authz.support.DLSRoleQueryValidator;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authz.store.CompositeRolesStore;
+
+import java.util.Set;
+
+public final class TransportUpdateApiKeyAction extends HandledTransportAction<UpdateApiKeyRequest, UpdateApiKeyResponse> {
+
+    private final ApiKeyService apiKeyService;
+    private final SecurityContext securityContext;
+    private final CompositeRolesStore rolesStore;
+    private final NamedXContentRegistry xContentRegistry;
+
+    @Inject
+    public TransportUpdateApiKeyAction(
+        final TransportService transportService,
+        final ActionFilters actionFilters,
+        final ApiKeyService apiKeyService,
+        final SecurityContext context,
+        final CompositeRolesStore rolesStore,
+        final NamedXContentRegistry xContentRegistry
+    ) {
+        super(UpdateApiKeyAction.NAME, transportService, actionFilters, UpdateApiKeyRequest::new);
+        this.apiKeyService = apiKeyService;
+        this.securityContext = context;
+        this.rolesStore = rolesStore;
+        this.xContentRegistry = xContentRegistry;
+    }
+
+    @Override
+    protected void doExecute(Task task, UpdateApiKeyRequest request, ActionListener<UpdateApiKeyResponse> listener) {
+        final var authentication = securityContext.getAuthentication();
+        if (authentication == null) {
+            listener.onFailure(new IllegalStateException("authentication is required"));
+            return;
+        } else if (authentication.isApiKey()) {
+            listener.onFailure(new IllegalArgumentException("authentication via an API key is not supported for updating API keys"));
+            return;
+        }
+
+        // Don't resolve and validate owner roles if the service is not enabled; this avoids a costly operation and also provides
+        // a clearer error message (service disabled error is prioritized over role descriptor validation errors)
+        apiKeyService.ensureEnabled();
```

Comment:
This block of code is quite similar to what we have in `ApiKeyGenerator`. Re-using it however would require refactoring around the `ApiKeyGenerator` class and I haven't come up with a good approach for that yet. One idea is to generalize `ApiKeyGenerator` to something like `ApiKeyManager` with `create` and `update` methods. Not convinced yet that's the way to go -- the class loses it's single responsibility that way.

I'm wondering around leaving the current implementation and revisiting a possible refactor. WYDT?

Reply:
I'd prefer to _not_ duplicating the code here. We had previous bugs in handling limited role descriptors. So my instinct would be reusing the logic in ApiKeyGenerator with some refactoring. The reusable part is mainly about getting the list of limited role descriptors for creating API keys (now updating as well). I think it can be refactored into its own method (still within ApiKeyGenerator) relatively easily, e.g.:

```java
public void generateApiKey(Authentication authentication, CreateApiKeyRequest request, ActionListener<CreateApiKeyResponse> listener) {
    if (authentication == null) {
        listener.onFailure(new ElasticsearchSecurityException("no authentication available to generate API key"));
        return;
    }
    apiKeyService.ensureEnabled();

    getUserRoleDescriptors(authentication, ActionListener.wrap(roleDescriptors -> {
        apiKeyService.createApiKey(authentication, request, roleDescriptors, listener);
    }, listener::onFailure));
}

private void getUserRoleDescriptors(Authentication authentication, ActionListener<Set<RoleDescriptor>> listener) {
    final ActionListener<Set<RoleDescriptor>> roleDescriptorsListener = ActionListener.wrap(roleDescriptors -> {
        for (RoleDescriptor rd : roleDescriptors) {
            try {
                DLSRoleQueryValidator.validateQueryField(rd.getIndicesPrivileges(), xContentRegistry);
            } catch (ElasticsearchException | IllegalArgumentException e) {
                listener.onFailure(e);
                return;
            }
        }
        listener.onResponse(roleDescriptors);
    }, listener::onFailure);

    final Subject effectiveSubject = authentication.getEffectiveSubject();

    // Retain current behaviour that User of an API key authentication has no roles
    if (effectiveSubject.getType() == Subject.Type.API_KEY) {
        roleDescriptorsListener.onResponse(Set.of());
        return;
    }

    rolesStore.getRoleDescriptorsList(effectiveSubject, ActionListener.wrap(roleDescriptorsList -> {
        assert roleDescriptorsList.size() == 1;
        roleDescriptorsListener.onResponse(roleDescriptorsList.iterator().next());

    }, roleDescriptorsListener::onFailure));
}
```

The above refactoring should not cause any cascading changes for both main and test code. One advantage of the separate and reusable method is that it can be easily moved to another place if necessary (e.g. if we want to merge it into ApiKeyService). I am not concerned about its single responsibility since I don't think we particularly wanted it. My guess is that it existed because we didn't want to expose RolesStore to ApiKeyService.

We might want to have more refactoring around ApiKeyService, either merging ApiKeyGenerator entirely into ApiKeyService, or as you mentioned, enhance ApiKeyGenerator to be something like ApiKeyManager as a high level facade of the Service class. In either case, we probably won't have this logic inside the transport action.

Reply:
I'm good with the refactoring 👍 I really like how focused and small `ApiKeyGenerator` is but I agree that the code re-use makes sense here.